Course navigation markup and styles

clrux opened the pull request:
This work addresses the style discrepancies after a positive markup update from @ahsan-ul-haq that increases the accessibility of the course navigation sidebar. The styles should closely match what existed before the markup updates.

Note: One goal was to send focus to the first menu item after a menu has opened. To do that, we had to remove the `slideUp()` and `slideDown()` functions, which slid open and closed the accordion panels, and replace them with `show()` and `hide()`. This is because after jQuery slides open the menu, it sends focus to the first item, but because the focus event is triggered before the animation completes, the first link is still hidden so focus is never sent.

__Browser testing:__

This code has been tested in the following browsers:
```
| Platform | Browser | Version | Passed | 
|----------|---------|---------|--------|
| Mac      | Chrome  | 43      | Yes    |
|          | Safari  | 8       | Yes    |
|          | Firefox | 24      | Yes    |
| Windows  | IE      | 10      | Yes    |
|          |         | 11      | Yes    |
|          | Firefox | 37      | Yes    |
```
---

@frrrances @marcotuts mind reviewing the FED and UI?

clrux commented:
test this please

clrux commented:
test this please

clrux commented:
test this please

clrux commented:
test this please

clrux commented:
test this please

clrux commented:
test this please

frrrances commented:
@clrux I did a quick pass on the interaction and it is looking good but I see two differences: 

1. the subsections are bold
2. the expand arrows don't expand when viewing another subsection: 

![screen shot 2015-05-27 at 9 36 18 am](https://cloud.githubusercontent.com/assets/4327102/7837292/2c55ce02-0454-11e5-9a48-d113e82a791c.png)

Please check those out. Otherwise, looks good! I will review code now.

frrrances commented:
@clrux I have reviewed your commits and have a few small comments. Once you address those, and the  interaction notes above, this should be ready for thumbs.

clrux commented:
Thanks @frrrances. I've gone through and made the tweaks you've requested and fixed the click area. Mind taking one more pass?

frrrances commented:
@clrux Code looks good. Thanks for making those changes! I am still seeing the arrow facing the wrong way when clicking around the sections though. Can you check it once more?

marcotuts commented on lms/static/sass/course/courseware/_sidebar.scss:
```diff
@@ -108,40 +100,46 @@
+            @extend %t-strong;
             margin-bottom: 0;
-            line-height: 1.3;
+            font-family: $sans-serif;
```
NIT: should our t-action extends also include a base font-family reset like copy / title extends? If so we could drop this line but probably not necessary here. Just a comment in passing.

marcotuts commented on lms/static/sass/course/courseware/_sidebar.scss:
```diff
@@ -168,19 +166,19 @@
 
           &:after {
+            @extend %t-regular;
+            @include transition(none);
```
NO-OP: the subtlest of transitions. The none transition. ;)

marcotuts commented:
Nothing I could see in a styling code review, :+1: (Note - I didn't review the python and other changes)

frrrances commented:
Aside from the blublublub, styles and interaction look good! The python needs a dev to review, and it would be good to have them look at the JS as well. 

(Once those things have been reviewed, and tests are passing, I give my :+1: )

clrux commented:
Thanks @frrrances. I've removed my test text and squashed into the previous commit. Who would you recommend to review the Python and Coffee/JS work, most of it performed by @ahsan-ul-haq.

clrux commented:
@cptvitamin and @ahsan-ul-haq I've made the necessary style changes. Please have someone review the Python and JS work here before merging. I'm going to re-assign the UX ticket back to @ahsan-ul-haq  to get this across the finish line. Please let me know if you need anything else here.

cptvitamin commented:
@clrux One small comment.  I noticed this because I have my default outlined cranked up in Firefox so it was very noticeable.  It was slightly noticeable in Chrome, but I think its an issue that should be addressed nonetheless.  In the existing sidebar nav, the padding on the ```<li>``` is evenly distributed top and bottom but on the the new design its all onthe bottom.  This cuts off the focus indicator on the top because the padding is zero there.
Existing:
![Existing padding](https://cloud.githubusercontent.com/assets/444368/7869771/edb035b6-0552-11e5-80dc-42c92f0bc5b0.png)

New:
![New Padding](https://cloud.githubusercontent.com/assets/444368/7869770/edaf498a-0552-11e5-938c-0dcf03af01e4.png)

cptvitamin commented:
Nice, thanks for addressing that issue so quickly, @clrux.  :+1: from me.   @symbolist and @cahrens can you give this a quick code review?  This work was started by @ahsan-ul-haq and @clrux did the styling.  Its basically a rebuild of the course nav (sidebar) front end code to make it more accessible.

cahrens commented:
I have created https://openedx.atlassian.net/browse/TNL-2281 for someone from TNL Cambridge to review. What is the timeframe that this PR needs to be merged?

clrux commented:
Thanks @cahrens. As soon as possible would be great as there are accessibility issues in the existing (live) version.

andy-armstrong commented:
@clrux I've been assigned to review this from T&L, and I'll be starting my review later today. Is there a sandbox that I can use?

andy-armstrong commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -157,6 +157,7 @@ def toc_for_course(request, course, active_chapter, active_section, field_data_c
         for chapter in chapters:
             # Only show required content, if there is required content
             # chapter.hide_from_toc is read-only (boo)
+            name_without_spaces = chapter.display_name_with_default.replace(" ", "-").replace(":", "")
```
There are all kinds of other characters that might cause problems, so I recommend using a more comprehensive solution. The T&L team is introducing a ``slugify`` function from Django 1.8 that I would recommend. Unfortunately it hasn't been merged yet, but you could add it in your PR too (or wait for https://github.com/edx/edx-platform/pull/8243 to land):

https://github.com/edx/edx-platform/blob/benmcmorran/team-api-a/common/djangoapps/util/model_utils.py#L154

clrux replied:
@andy-armstrong I would tag @ahsan-ul-haq on the functional bits. I only handled styling for this piece of work.

I will, however, get you a sandbox spun up!

ahsan-ul-haq replied:
@andy-armstrong I have looked at the ```slugify``` function that T&L team is trying to implement and it looks great to me. So I will follow up the PR and use that once https://github.com/edx/edx-platform/pull/8243 is merged.

andy-armstrong commented:
There's an issue where you can get two selected items in the accordion:
1. Navigate to a unit that isn't the first in a subsection, 
e.g. http://clrux-2101.m.sandbox.edx.org/courses/edX/DemoX.1/2015/courseware/0af8db2309474971bfa70cda98668a30/9c1aacbb2795470e8473b059b59c3344/
2. Click on the subsection and then both the original and the first units will be shown as selected.

![image](https://cloud.githubusercontent.com/assets/5985072/7947411/f69c66ee-094a-11e5-9a44-bc490d3ac836.png)

I verified that this doesn't happen on prod.

andy-armstrong commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -182,6 +183,7 @@ def toc_for_course(request, course, active_chapter, active_section, field_data_c
                                      })
             toc_chapters.append({
                 'display_name': chapter.display_name_with_default,
+                'display_id': name_without_spaces,
```
@ahsan-ul-haq I would call the variable ``display_id`` since it isn't a name without spaces anymore.

clrux replied:
@andy-armstrong This isn't a bug. Some accessibility work that was done, sends focus to the first item when the menu expands. So what you're seeing is the first item with focus (I imagine you just opened the menu) and the second which is active.

frrrances replied:
I also noticed what @andy-armstrong mentioned and realized what was going on with the focus and selected, but it might be worth revisiting the focus state to be a bit more different than the selected state so it  doesn't look weird, or have the js focus on the selected item instead.

clrux replied:
I personally like the default (blue glow) focus state. It's being overridden somewhere in the platform to be dotted lines, which is why it's showing up here funky. I'm thinking maybe we remove the gray background on `focus` which would remove the confusion. Thoughts?

frrrances replied:
Removing the gray background makes sense to me.

andy-armstrong commented:
@ahsan-ul-haq I'm done with my review. The code looks great and the functionality seems good. Tag me when my feedback has been addressed and I'll do a second review.

ahsan-ul-haq commented:
@clrux I will take this PR from here and address feedback from andy

clrux commented:
@ahsan-ul-haq Thanks, please let me know if you need anything on my end.

clrux commented:
@andy-armstrong @frrrances I just pushed a small update that makes an adjustment to the `focus` style.

clrux commented:
@ahsan-ul-haq How is this work going? Do you need anything from me?

clrux commented:
@ahsan-ul-haq Continued this on a new branch. Closing this one.
